### PR TITLE
Fix OOM by empty stack of exception.

### PR DESCRIPTION
**Pull request by wu-sheng**
First, I provide more readable codes for #2926 fix.
Then, as people are asking for bootstrap class enhancing, so I provided this as an optional mechanism. I am still hesitating whether we should accept plugin for bootstrap class, such as `java.net.URLConnection`

@ascrutae @candyleer We should test by using this to enhance some class in `rt.jar`. Please verify.

### CODE REVIEW

**Comment by wu-sheng** [comment]
@IanCao @zhaoyuguang @JaredTan95 I hope you have time to verify.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java [hunk 1]
```diff
@@ -30,10 +32,12 @@
...
         final StringBuilder stackMessage = new StringBuilder();
         Throwable causeException = throwable;
-        while (causeException != null) {
+
+        int depth = CAUSE_EXCEPTION_DEPTH;
+        while (causeException != null || depth == 0) {
             stackMessage.append(printExceptionInfo(causeException));
 
-            boolean overMaxLength = printStackElement(throwable.getStackTrace(), new AppendListener() {
+            boolean isLookDeeper = printStackElement(causeException.getStackTrace(), new AppendListener() {
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: @Arugal @SummerOfServenteen I think there are two places originally wrong, which also trigger #2926

Thank @Arugal to catch this.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java [hunk 2]
```diff
@@ -43,11 +47,12 @@ public boolean overMaxLength() {
                 }
             });
 
-            if (overMaxLength) {
+            if (isLookDeeper) {
                 break;
             }
 
-            causeException = throwable.getCause();
+            causeException = causeException.getCause();
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: The other one triggers #2926

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: FYI @ascrutae This is the root cause and being here for a long time. `OverSize` check avoids OOM in most cases only.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java (hunk 1 above)
Comment (SummerOfServenteen) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: maybe depth != 0

Comment (SummerOfServenteen) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: maybe ( causeException != null  && depth != 0)

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java]: Yes. Done.

**Comment by wu-sheng** [comment]
Bootstrap loader is more complex than this. I will remove it from this PR.

**Pull request merged**